pkg/plugins/generic/generic_plugin.go

```diff
@@ -124,6 +124,31 @@ func (p *GenericPlugin) OnNodeStateChange(new *sriovnetworkv1.SriovNetworkNodeSt
 	return
 }
 
+// OnNodeStatusChange verify whether SriovNetworkNodeState CR status present changes on configured VFs.
+func (p *GenericPlugin) OnNodeStatusChange(new *sriovnetworkv1.SriovNetworkNodeState) bool {
```

Comment:
I see no need to create this new interface method on the plugin interface: there is only one implementation, explicitly invoked on the generic instance. What about moving this method next to `utils.NeedUpdate(...)`?

Reply:
That's what I was planning to do initially, but like you said, this is something specific to the generic plugin. It is wrong to expose it generally. For instance, when the platform type is VirtualOpenstack, we don't load the generic plugin at all. This also provides a framework for other plugins to perform their own implementation when status changes. I see that Mellanox also inspects the interfaces or perhaps something in the virtual plugin. That can be added in the future.

Reply:
I haven't understood the full PR yet, but one comment is that **IF** we decide to add `OnNodeStatusChange` function, we need a much clearer name that is has more than 2 characters differing from `OnNodeStateChange`

Reply:
The main idea of the pr is to start the reconciliation when an user externally changes a parameter of a VF which was set via the operator with a sriov policy. My argument is that the operator should not only act when the user changes the spec but also when the user externally modifies something that affects what was set in the spec.

For instance, we have a customer that was setting the mtu to 9000 with a sriov policy. After that, they were manually changing the mtu of some vfs. Nothing happened. The status of the `SriovNetworkNodeState`  object detected the change but the reconciliation was not triggered because we only do when the generation changes (spec changes). However, this became an issue when the resource version of the device config map, which is part of that spec, weekly changed due to a bug on the kube-api server. The reconciliation loop started and we could see 2 issues.

1.  `OnNodeStateChange` of the generic plugin returned that a drain was needed because the spec and status didn't match. The customer was complaining of this unexpected draining. Unexpected from their side.
2. `Apply` of the generic plugin never reconcile this discrepancy because its logic compares the current spec with the spec of the previously applied `SriovNetworkNodeState` object. Due to this, user was experiencing a weekly draining of nodes (the api-server bug happened every 7 days). We shouldn't rely on a save object to decide when to apply changes. This goes against the logic of `OnNodeStateChange` and behaves differently when the node restarts (saved data is lost).

This pr is addressing all this. Perhaps, `CheckStatusChanges` is better. What do you think @bn222 ?

Reply:
if the main idea is to detect when a parameter changes, then I think we need to properly address this instead of only checking when reconciling. I'm in favour of snooping netlink messages to intercept whenever mtu (or other things) change and react immediately. Waiting until reconcile happens makes things a lot more difficult to debug.

The new proposed name is in any case better, but my above comment should be address too I think.

Reply:
I don't agree. That would add more complexity for something that we don't expect users do often. Besides, we don't have the need to react immediately.

Just to be clear, the idea is not act only when reconciliation happens. I'm relying on the `Run` method of `NodeStateStatusWriter` which runs `pollNicStatus` and updates the status of the `SriovNetworkNodeState` every 30 seconds. So, at the worst case, the reconciliation would happen after 30 seconds of making a change.

I think this is good enough for our use cases. No need to add more complexity. This is a good start. Perhaps if some day a user wants their externally changes be reverted immediately, then we can explore other alternatives, such as snooping netlink messages as you said.

Reply:
How do you address debuggability today? Having snooping in place would allow us to put a message like "Detected user changing xyz but policy is in place to mange xyz so changing it back", with a timestamp.

Reply:
First of all, we don't refer to policies at all in the daemon. Everything goes around the `SriovNetworkNodeState` object. Regarding debuggability, note that the current logic is to only start the reconciliation loop when the generation field of `SriovNetworkNodeState` has changed. So, when it hasn't changed, the idea is to run `OnNodeStatusChange` (soon to be `CheckStatusChanges`). If changes are detected, there will be logs like so:

log.Log.Info("OnNodeStatusChange(): status changed for interface", "address", iface.PciAddress)

So, there are info for debuggability.

Reply:
changed the name to `CheckStatusChanges`